README.md
# Editmode Standalone

Editmode allows you to turn plain text in your website into easily inline-editable bits of content that can be managed by anyone with no technical knowledge.

## Installation
**From CDN:** Add the following script to the end of your `<head>` section.
```HTML
<script src="https://unpkg.com/editmode-standalone@^1/dist/standalone.js"></script>
```

**Set an Editmode Project ID:** Add the following script to the end of your `<head>` or `<body>` section.
```HTML
<script>
  EditmodeStandAlone.projectId = "YOUR-PROJECT-ID"
</script>
```
*Don't have a project id? Sign up for one [here](https://editmode.com/users/sign_up?private_beta=true)*


## Usage

*Single Content*
```HTML
<p chunk-id="some-chunk-identifier">I'm a placeholder</p>

<img chunk-id="some-img-chunk-identifier"  src="https://backup-img-src.png"/>
```

*Collection*
```HTML
<template collection-id="some-collection-identifier" class="container-class" itemClass="classForEachItems">
  <p field-id="field-id-or-name">Placeholder for this field</p>

  <img field-id="field-id-or-name"/>
</template>
```

*Filter Collection*

Use `tags` attribute to filter collection items by tags, use `limit` attribute to limit how many collection items will be rendered.

Seperate the tags with space for multiple tags `tags="heropost recentposts"`

```HTML
<template collection-id="some-collection-identifier" tags="recentposts" limit="3">
  <p field-id="field-id-or-name">Placeholder for this field</p>

  <img field-id="field-id-or-name"/>
</template>
```


*Variable Parsing*
```HTML
<p variables="{name: 'Jen'}" chunk-id="some-chunk-identifier">Hi, User</p>
```

*Working with Image Transformation*

Use `transform` attribute to perform real-time image transformations to deliver perfect images to the end-users.

This chunk should render an image with 200 x 200 dimension:
```HTML
<img chunk-id="some-img-chunk-identifier"  transform="w-200 h-200 c-force" />
```
Please see complete list of [transformation parameters](https://editmode.com/docs#/imagekit_properties)


### Demo
[Open in Repl.it](https://repl.it/@jengkarlong/EditmodeStandAlone-Example)

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/puuripurii"><img src="https://avatars.githubusercontent.com/u/26903002?v=4?s=40" width="40px;" alt=""/><br /><sub><b>Jen Villaganas </b></sub></a><br /><a href="https://github.com/Editmodelabs/editmode-standalone/commits?author=puuripurii" title="Code">💻</a> <a href="https://github.com/Editmodelabs/editmode-standalone/commits?author=puuripurii" title="Documentation">📖</a> <a href="#infra-puuripurii" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="https://github.com/Editmodelabs/editmode-standalone/commits?author=puuripurii" title="Tests">⚠️</a></td>
    <td align="center"><a href="https://github.com/jenvillaganas"><img src="https://avatars.githubusercontent.com/u/26903002?v=4?s=40" width="40px;" alt=""/><br /><sub><b>Jen Villaganas </b></sub></a><br /><a href="https://github.com/Editmodelabs/editmode-standalone/commits?author=jenvillaganas" title="Code">💻</a> <a href="https://github.com/Editmodelabs/editmode-standalone/commits?author=jenvillaganas" title="Documentation">📖</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

CHANGELOG.md
# v1.0.10 (Mon Sep 06 2021)

#### ⚠️ Pushed to `master`

- Replace new powered by logo ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.9 (Mon Sep 06 2021)

#### 🐛 Bug Fix

- New Magic Editor Compatibility [#6](https://github.com/editmodelabs/editmode-standalone/pull/6) ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.8 (Wed Jun 02 2021)

#### ⚠️ Pushed to `master`

- Add option to override magic editor url ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.7 (Mon Mar 22 2021)

#### 🐛 Bug Fix

- Branch, Collection Filters, Default Chunks [#5](https://github.com/editmodelabs/editmode-standalone/pull/5) ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.6 (Sat Mar 13 2021)

#### 🐛 Bug Fix

- Handle Transform Image [#4](https://github.com/editmodelabs/editmode-standalone/pull/4) ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.5 (Sat Mar 13 2021)

#### ⚠️ Pushed to `master`

- Merge branch 'master' of github.com:editmodelabs/editmode-standalone ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.4 (Sat Mar 13 2021)

#### 🐛 Bug Fix

- Add Store and Get Cache [#3](https://github.com/editmodelabs/editmode-standalone/pull/3) ([@jenvillaganas](https://github.com/jenvillaganas))

#### ⚠️ Pushed to `master`

- build file ([@jenvillaganas](https://github.com/jenvillaganas))

#### Authors: 1

- Jen Villaganas  ([@jenvillaganas](https://github.com/jenvillaganas))

---

# v1.0.3 (Sun Mar 07 2021)

#### 🐛 Bug Fix

- Parse Variable [#2](https://github.com/editmodelabs/editmode-standalone/pull/2) ([@puuripurii](https://github.com/puuripurii))

#### Authors: 1

- Jen Villaganas  ([@puuripurii](https://github.com/puuripurii))

---

# v1.0.2 (Sat Mar 06 2021)

:tada: This release contains work from a new contributor! :tada:

Thank you, Jen Villaganas  ([@puuripurii](https://github.com/puuripurii)), for all your work!

#### 🐛 Bug Fix

- Use Template Tag for Collection Render [#1](https://github.com/editmodelabs/editmode-standalone/pull/1) ([@puuripurii](https://github.com/puuripurii))

#### Authors: 1

- Jen Villaganas  ([@puuripurii](https://github.com/puuripurii))

---

# v1.0.1 (Sun Feb 28 2021)

:tada: This release contains work from a new contributor! :tada:

Thank you, Jen Villaganas  ([@puuripurii](https://github.com/puuripurii)), for all your work!

#### ⚠️ Pushed to `master`

- create separate config for dev ([@puuripurii](https://github.com/puuripurii))
- use rollup instead of webpack ([@puuripurii](https://github.com/puuripurii))
- configure webpack ([@puuripurii](https://github.com/puuripurii))
- Merge branch 'main' of github.com:editmodelabs/editmode-standalone into main ([@puuripurii](https://github.com/puuripurii))
- add auto and all contributores ([@puuripurii](https://github.com/puuripurii))
- Create release.yml ([@puuripurii](https://github.com/puuripurii))
- Update readme ([@puuripurii](https://github.com/puuripurii))
- Add readme, add collection placeholder ([@puuripurii](https://github.com/puuripurii))
- Delete .DS_Store ([@puuripurii](https://github.com/puuripurii))
- Delete node_modules directory ([@puuripurii](https://github.com/puuripurii))
- include dist folder ([@puuripurii](https://github.com/puuripurii))
- ignore DS_STORE ([@puuripurii](https://github.com/puuripurii))
- update gitigonr ([@puuripurii](https://github.com/puuripurii))
- add gitignore ([@puuripurii](https://github.com/puuripurii))
- Render chunks and collections ([@puuripurii](https://github.com/puuripurii))

#### Authors: 1

- Jen Villaganas  ([@puuripurii](https://github.com/puuripurii))


src/utils.js
// Thanks @stimulus:
// https://github.com/stimulusjs/stimulus/blob/master/packages/%40stimulus/core/src/application.ts
export function domReady() {
    return new Promise(resolve => {
        if (document.readyState == "loading") {
            document.addEventListener("DOMContentLoaded", resolve)
        } else {
            resolve()
        }
    })
}

export async function api(path, config = {}) {
    const { parameters } = config
    const url = new URL("https://api.editmode.com")
    const urlparams = url.searchParams

    url.pathname = path

    if (parameters) {
        for (let [key, val] of Object.entries(parameters)) {
            if (val) {
                if (Array.isArray(val)) {
                    val.forEach(v => urlparams.append(key + "[]", v))
                } else {
                    urlparams.append(key, val)
                }
            }
        }
    }

    return fetch(url.toString(),
        {
            method: 'get',
            headers: {
                Accept: "application/json",
                referrer: window.location.href
            }
        }
    ).then(response => response.json())
}

export const parseVariable = (content, variables) => {
    if (variables) {
        try {
            eval(`variables = ${variables}`)
            if (typeof variables != 'object') throw "Invalid Variable Value " + variables
        } catch (error) {
            console.log(error)
        }

        const tokens = content && (content.match(/\{{(.*?)\}}/g) || []).map(t => t.substr(2, t.length - 4))

        if (tokens) {
            tokens.forEach(function (token) {
                const value = variables && variables[token] || ""
                const emVar = `<em-var data-chunk-variable="${token}" data-chunk-variable-value="${value}">${value}</em-var>`
                content = content.replace(`{{${token}}}`, emVar);
            });
        }
    }

    return content
}

export const setBranchId = () => {
    const url = new URL(window.location.href)
    const id = url.searchParams.get('em_branch_id')

    return id || ""
}

export const getCachedData = (id) => {
    const data = localStorage.getItem(id)
    if (data) {
        return JSON.parse(data)
    }
}

export const storeCache = (id, data) => {
    if (!data) return
    localStorage.setItem(id, JSON.stringify(data))
}

export const setTransformAttributes = (url, transform) => {
    try {
        transform = transform.trim().replace(/\s+/g, ' ').replaceAll(" ", ",")
        console.log(transform)
        let newUrl = new URL(url)

        newUrl.searchParams.append('tr', transform)

        return newUrl.toString()
    } catch (er) {
        console.warn(er)
        return url
    }
}

export const logger = (message) => {
    // TODO: Add type to change color between warn info and danger
    console.log(`%c ${message}`, 'color: #bada55');
}


export const storeTimedCache = (id, data) => {
    const expiry = new Date(new Date().setHours(new Date().getHours() + 1)); // Set 1 hour expiration
    const item = {
        value: data,
        expiry: expiry.getTime(),
    };
    localStorage.setItem(id, JSON.stringify(item));
};

export const getTimedCachedData = (id) => {
    const cachedItem = localStorage.getItem(id);
    if (!cachedItem) return null;
    const item = JSON.parse(cachedItem);
    const now = new Date();
    if (now.getTime() > item.expiry) {
        localStorage.removeItem(id);
        return null;
    }
    return item.value;
};

src/components.js
import { parseVariable, setTransformAttributes } from './utils'

const Component = {
  renderChunk: function(el, chunk, collectionItem = null, dummy = false) {
    if (typeof chunk.content == 'undefined') return el

    const isNotEditable = el.getAttribute('editable') === false
    const transform = el.getAttribute('transform')
    let variables = el.getAttribute('variables')

    el.dataset.chunk = chunk.identifier
    el.dataset.chunkType = chunk.chunk_type
    el.dataset.chunkEditable = !isNotEditable

    if (collectionItem) {
      el.dataset.parentIdentifier = collectionItem.identifier
      el.dataset.customFieldIdentifier = chunk.custom_field_identifier
    }

    if (chunk.chunk_type == 'rich_text') {
      el.classList.add("editmode-richtext-editor")
    }

    if (chunk.chunk_type == 'image') {
      if (dummy) {
        const svgImage = this.createImagePlaceholder(el)
        el = svgImage
      } else {
        if (chunk.content && transform) {
          chunk.content = setTransformAttributes(chunk.content, transform)
        }
        el.src = chunk.content
      }

    } else {
      el.innerHTML = parseVariable(chunk.content, variables)
    }

    return el
  },

  renderCollection: function(el, collectionItems, limit = null, tags = []) {
    const collectionContainer = document.createElement('div')
    const containerClass = el.getAttribute('class') || ""

    // Limit item display
    if (limit) collectionItems = collectionItems.splice(0, limit)
    // Add tags in dataset
    if (tags.length) collectionContainer.dataset.chunkTags = tags.join("|")

    if (collectionItems.length) {
      collectionContainer.setAttribute('class', `${containerClass} chunks-collection-wrapper`)
      collectionContainer.dataset.chunkCollectionIdentifier = collectionItems[0].collection.identifier
      console.log(tags)

      collectionItems.forEach(collectionItem => {
        const itemWrapper = this.createCollectionItem(el, collectionItem)
        collectionContainer.appendChild(itemWrapper)
      })

      // Info: We add hidden item to every collection so we can use
      // this el as template for a new collection item 
      const dummyItem = this.createCollectionItem(el, collectionItems[0], true)
      
      collectionContainer.appendChild(dummyItem)
      el.replaceWith(collectionContainer)
    } else {
      const emptyCollectionEl = document.createElement('div')
      emptyCollectionEl.style.display = "none"
      emptyCollectionEl.innerText = "No Collection Item Found"
      el.replaceWith(emptyCollectionEl)
      console.warn("No collection item found for collection_id: " + el.getAttribute('collection-id'))
    }
  },

  createImagePlaceholder: function(originalEl) {
    const svgImage = document.createElement('svg')
    svgImage.classList = originalEl.classList.value
    svgImage.setAttribute('fill', 'currentColor')
    svgImage.setAttribute('viewBox', "0 0 20 20")
    svgImage.innerHTML = '<path fill-rule="evenodd" d="M4 3a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V5a2 2 0 00-2-2H4zm12 12H4l4-8 3 6 2-4 3 6z" clip-rule="evenodd"></path>'
    Object.assign(svgImage.dataset, originalEl.dataset);
    return svgImage
  },

  createCollectionItem: function(el, item, dummy = false) {
    let template = el.innerHTML
    const fields = el.content.querySelectorAll("[field-id]")
    let itemWrapper = document.createElement('div')
    const itemClass = el.getAttribute('itemClass') || ""

    itemWrapper.setAttribute('class', itemClass)
    
    if (dummy) {
      itemWrapper = document.createElement('em-template')
      // Create Dummy Item for new collection item
      itemWrapper.classList.add("chunks-col-placeholder-wrapper")
      itemWrapper.style.display = "none"
    }else{
      itemWrapper.classList.add("chunks-collection-item--wrapper")
    }

    const fieldChunks = item.content

    fields.forEach(fieldTemplate => {
      const fieldChunkData = fieldChunks.find(fieldChunk => fieldTemplate.getAttribute('field-id') === fieldChunk.custom_field_identifier || fieldTemplate.getAttribute('field-id') === fieldChunk.custom_field_name) || {}

      if (dummy) fieldChunkData.content = "" // Info: Line 42

      const cloneFieldElement = fieldTemplate.cloneNode(true)

      const newComponent = this.renderChunk(cloneFieldElement, fieldChunkData, item, dummy) || fieldTemplate
      template = template.replace(fieldTemplate.outerHTML, newComponent.outerHTML)
    })

    itemWrapper.innerHTML = template

    return itemWrapper
  }

}

export default Component

src/index.js
import { 
  domReady, api, getCachedData, storeCache,
  setBranchId, logger, getTimedCachedData, storeTimedCache
} from './utils'
import Component from "./components"

const EditmodeStandAlone = {
  projectId: null,
  branchId: setBranchId(),
  defaultChunks: [],

  // Initialize
  start: async function () {
    console.log('Discovering contents...')

    await domReady() 

    this.attachWatermark()
    this.discoverContents()
    this.addMagicEditorPlugin()

    window.chunksProjectIdentifier = this.projectId
  },

  attachWatermark: function() {
    const div = document.createElement('div')
    div.innerHTML = `<div style="z-index: 9999; bottom: 8px; right: 8px; position: fixed; opacity: 1; display: flex; align-items: center; background: rgba(255, 255, 255, 0.9); cursor: pointer; border-radius: 5px; padding: 2px 5px 2px 2px; box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px, rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;"><svg width="20" height="27" style="margin-right: 5px;"  viewBox="0 0 57 64" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M4.47575 11.9681C4.47575 11.9681 22.7685 1.26837 24.5608 0.266941C26.3531 -0.734485 27.9296 1.26837 27.9296 3.17864C27.9296 5.08891 27.9296 21.2701 27.9296 26.1723C27.9296 31.0745 27.3481 31.0745 25.4907 32.3321C23.6332 33.5898 4.35396 44.8099 3.29458 45.3859C2.23519 45.9619 0 45.1025 0 42.5066C0 39.9107 0 23.0363 0 18.6204C0 14.2045 4.47575 11.9681 4.47575 11.9681Z" fill="#203260"/>
    <mask id="mask0" mask-type="alpha" maskUnits="userSpaceOnUse" x="0" y="0" width="28" height="46">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M4.47575 11.9681C4.47575 11.9681 22.7685 1.26837 24.5608 0.266941C26.3531 -0.734485 27.9296 1.26837 27.9296 3.17864C27.9296 5.08891 27.9296 21.2701 27.9296 26.1723C27.9296 31.0745 27.3481 31.0745 25.4907 32.3321C23.6332 33.5898 4.35396 44.8099 3.29458 45.3859C2.23519 45.9619 0 45.1025 0 42.5066C0 39.9107 0 23.0363 0 18.6204C0 14.2045 4.47575 11.9681 4.47575 11.9681Z" fill="white"/>
    </mask>
    <g mask="url(#mask0)">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.76054 22.4622C7.76054 22.4622 26.0533 11.7625 27.8456 10.7611C29.6379 9.75966 31.2144 11.7625 31.2144 13.6728C31.2144 15.583 31.2144 31.7643 31.2144 36.6664C31.2144 41.5686 30.6329 41.5686 28.7755 42.8263C26.918 44.084 7.63875 55.304 6.57937 55.88C5.51998 56.456 3.28479 55.5966 3.28479 53.0007C3.28479 50.4048 3.28479 33.5304 3.28479 29.1145C3.28479 24.6986 7.76054 22.4622 7.76054 22.4622Z" fill="#000719"/>
    </g>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M18.4527 20.7957C18.4527 20.7957 36.7455 10.0959 38.5378 9.09452C40.3301 8.09309 41.9066 10.0959 41.9066 12.0062C41.9066 13.9165 41.9066 30.0977 41.9066 34.9999C41.9066 39.902 41.3251 39.902 39.4677 41.1597C37.6102 42.4174 18.331 53.6374 17.2716 54.2135C16.2122 54.7895 13.977 53.9301 13.977 51.3341C13.977 48.7382 13.977 31.8638 13.977 27.448C13.977 23.0321 18.4527 20.7957 18.4527 20.7957Z" fill="#223464"/>
    <mask id="mask1" mask-type="alpha" maskUnits="userSpaceOnUse" x="13" y="8" width="29" height="47">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M18.4527 20.7957C18.4527 20.7957 36.7455 10.0959 38.5378 9.09452C40.3301 8.09309 41.9066 10.0959 41.9066 12.0062C41.9066 13.9165 41.9066 30.0977 41.9066 34.9999C41.9066 39.902 41.3251 39.902 39.4677 41.1597C37.6102 42.4174 18.331 53.6374 17.2716 54.2135C16.2122 54.7895 13.977 53.9301 13.977 51.3341C13.977 48.7382 13.977 31.8638 13.977 27.448C13.977 23.0321 18.4527 20.7957 18.4527 20.7957Z" fill="white"/>
    </mask>
    <g mask="url(#mask1)">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M22.3409 31.4244C22.3409 31.4244 40.6337 20.7246 42.426 19.7232C44.2183 18.7218 45.7947 20.7246 45.7947 22.6349C45.7947 24.5452 45.7947 40.7264 45.7947 45.6286C45.7947 50.5307 45.2132 50.5307 43.3558 51.7884C41.4984 53.0461 22.2191 64.2661 21.1597 64.8422C20.1003 65.4182 17.8651 64.5587 17.8651 61.9628C17.8651 59.3669 17.8651 42.4925 17.8651 38.0767C17.8651 33.6608 22.3409 31.4244 22.3409 31.4244Z" fill="#000719"/>
    </g>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M33.1693 30.915C33.1693 30.915 51.4781 20.3758 53.2719 19.3894C55.0658 18.403 56.6437 20.3758 56.6437 22.2574C56.6437 24.139 56.6437 40.0775 56.6437 44.9062C56.6437 49.7348 56.0616 49.7348 54.2026 50.9736C52.3435 52.2124 33.0474 63.2642 31.9871 63.8316C30.9268 64.3989 28.6896 63.5524 28.6896 60.9954C28.6896 58.4385 28.6896 41.8172 28.6896 37.4675C28.6896 33.1179 33.1693 30.915 33.1693 30.915Z" fill="#405489"/>
    </svg>
    <span style="font-size: 13px; font-weight: 600;">Powered by Editmode</span></div>`

    const cacheId =  this.projectId + "_provider"
    let project = getTimedCachedData(cacheId)

    if (!project) {
      api(`/projects/${this.projectId}`)
          .then((res) => {
            storeTimedCache(cacheId, res.data);
            project = res
            if ( project.has_watermark ) {
              document.body.append(div)
            }
          })
          .catch((error) => {
            console.error(error);
          });
    } else {
      if ( project.has_watermark ) {
        document.body.append(div)
      }
    }
  },

  discoverContents: function() {
    this.discoverChunks(el => {
      this.getChunk(el)
    })

    this.discoverCollections(el => {
      this.getCollection(el)
    })
  },

  // Scan DOM Search for tags with chunk-id attr
  discoverChunks: function(callback) {
    const chunkElements = document.querySelectorAll('[chunk-id]');

    chunkElements.forEach(el => {
      callback(el)
    })
  },

  // Get chunk data from Editmode API
  getChunk: function(el) {
    const chunkId = el.getAttribute('chunk-id')
    const chunkProjectId = el.getAttribute('project-id')
    const cacheid = chunkId + this.projectId + this.branchId
    const cacheData = getCachedData(cacheid)
    
    if (cacheData) {
      logger('Rendering from cache: ' + chunkId)
      Component.renderChunk(el, cacheData)
    }

    // Render from default chunk
    if (!cacheData && this.defaultChunks.length) {
      logger('Rendering from defaultChunks: ' + chunkId)
      let chunk = this.defaultChunks.find(c => c.identifier == chunkId )
      storeCache(cacheid, chunk)
      Component.renderChunk(el, chunk)
    }

    api(`/chunks/${chunkId}`, 
      { 
        parameters: { 
          project_id: chunkProjectId || this.projectId,
          branch_id: this.branchId
        }
      }
    ).then(chunk => {
      storeCache(cacheid, chunk)

      // If no cache data, render content from API
      if (!cacheData && !this.defaultChunks.length) {
        logger('Rendering from API: ' + chunkId)
        Component.renderChunk(el, chunk)
      }
    })
  },

  // Scan DOM Search for collection tags
  discoverCollections: function(callback) {
    const collectionElements = document.querySelectorAll('template[collection-id]')

    collectionElements.forEach(el => {
      callback(el)
    })
  },

   // Get collection data from Editmode API
  getCollection: function(el) {
    const collectionId = el.getAttribute('collection-id')
    const chunkProjectId = el.getAttribute('project-id')
    const limit = el.getAttribute('limit')
    let tags = el.getAttribute('tags') || ""
    tags = tags.split(" ").filter(Boolean) // Filter boolean will remove empty strings

    const cacheid = collectionId + this.projectId + this.branchId + tags.join("")
    const cacheData = getCachedData(cacheid)

    if (cacheData) {
      logger('Rendering from cache: ' + collectionId)
      Component.renderCollection(el, cacheData, limit, tags)
    }

    if (!tags.length && !cacheData && this.defaultChunks.length) {
      logger('Rendering from defaultChunks: ' + collectionId)
      let chunks = this.defaultChunks.filter(c => c.collection && c.collection.identifier == collectionId)
      storeCache(cacheid, chunks)
      Component.renderCollection(el, chunks, limit, tags)
    }

    api("/chunks/", 
      {
        parameters: {
          collection_identifier: collectionId,
          project_id: chunkProjectId || this.projectId,
          branch_id: this.branchId,
          tags: tags
        }
      }).then(res => {
        let chunks = res.chunks
        storeCache(cacheid, chunks)

        if (!cacheData && !this.defaultChunks.length) {
          logger('Rendering from defaultChunks: ' + collectionId)
          Component.renderCollection(el, chunks, limit, tags)
        }
      })
  },

  // Add magic editor plugin script tag before closing body tag
  addMagicEditorPlugin: function() {
    let s = document.createElement('script')
    let url =  "https://unpkg.com/editmode-magic-editor@~1/dist/magic-editor.js"
    if(window.editmodeMagicEditorUrl) url = window.editmodeMagicEditorUrl
    s.setAttribute('src', url)
    s.async = true

    document.body.appendChild( s )
  }
}


window.EditmodeStandAlone = EditmodeStandAlone
window.EditmodeStandAlone.start()

export default EditmodeStandAlone
